dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckDataRelationsValidationHook.java

```diff
@@ -434,18 +428,9 @@ private boolean trackedEntityTypesMatch( TrackerBundle bundle, Program program,
                 .equals( trackedEntityInstance.getTrackedEntityType().getUid() );
         }
 
-        final Optional<ReferenceTrackerEntity> reference = bundle.getPreheat()
-            .getReference( enrollment.getTrackedEntity() );
-        if ( reference.isPresent() )
-        {
-            final Optional<TrackedEntity> tei = bundle.findTrackedEntityByUid( enrollment.getTrackedEntity() );
-            if ( tei.isPresent() )
-            {
-                return tei.get().getTrackedEntityType().isEqualTo( program.getTrackedEntityType() );
-            }
-        }
+        final Optional<TrackedEntity> tei = bundle.findTrackedEntityByUid( enrollment.getTrackedEntity() );
+        return tei.map( te -> te.getTrackedEntityType().isEqualTo( program.getTrackedEntityType() ) ).orElse( false );
```

Comment:
I think you can inline `tei`. Using some indentation/newlines it should be readable.